chain/chain/src/types.rs

```diff
@@ -557,19 +557,35 @@ pub trait RuntimeAdapter: Send + Sync {
     ) -> Result<Ordering, Error>;
 
     /// Build receipts hashes.
-    fn build_receipts_hashes(&self, receipts: &[Receipt]) -> Vec<CryptoHash> {
-        let mut receipts_hashes = vec![];
-        for shard_id in 0..self.num_shards() {
-            // importance to save the same order while filtering
-            let shard_receipts: Vec<Receipt> = receipts
-                .iter()
-                .filter(|&receipt| self.account_id_to_shard_id(&receipt.receiver_id) == shard_id)
-                .cloned()
-                .collect();
-            receipts_hashes
-                .push(hash(&ReceiptList(shard_id, &shard_receipts).try_to_vec().unwrap()));
+    // Due to borsh serialization constraints, we have to use `&Vec<Receipt>` instead of `&[Receipt]`
+    // here.
+    fn build_receipts_hashes(&self, receipts: &Vec<Receipt>) -> Vec<CryptoHash> {
+        if self.num_shards() == 1 {
+            return vec![hash(&ReceiptList(0, receipts).try_to_vec().unwrap())];
         }
-        receipts_hashes
+        let mut account_id_to_shard_id = HashMap::new();
+        let mut shard_receipts = HashMap::new();
```

Comment:
What if `shard_receipts` were a `Vec` from the start? Say 

```rust
let mut shard_receipts = (0..self.num_shards()).map(|_| Vec::new()).collect();
```

Then the loop at the bottom would simply be

```
shard_receipts.into_iter().enumerate().map(|(rs, i)| {
    let bytes = (i, rs).try_to_vec().unwrap();
    hash(&bytes)
}).collect()
```

which might be faster than going through the `HashMap`.

Reply:
This change to `Vec` has been made in https://github.com/nearprotocol/nearcore/pull/3235/commits/4481e964c205cb0f943bb6012edd15ed61dcd28d